src/box/box.cc

```diff
@@ -3106,8 +3106,17 @@ box_demote(void)
 		if (raft->state != RAFT_STATE_LEADER)
 			return 0;
 	}
-	if (box_trigger_elections() != 0)
-		return -1;
+	bool is_cfg_candidate_saved = raft->is_cfg_candidate;
```

Comment:
The problem here, that if quorum is 1, the node won't become leader automatically.

```
box.cfg{election_mode = 'candidate'}
box.ctl.demote()
-- Node remains in "follower" state until box.ctl.promote is called.
```

Before this patch Tarantool automatically calls `raft_sm_become_leader` from `raft_worker_handle_io`. However, with this patch the node is not a candidate at all, so it doesn't transition to leader. Moreover, as timer is not started, new term won't happen at all. The node is in "follower" until `box.ctl.promote` is called. This is definitely not a correct behavior.

Reply:
I also don't really like, that we make raft transit between candidate <-> not candidate state during demote a lot. This causes timer to be restarted. The solution with new internal flag in `raft` may be better. Up to you

Reply:
Reworked the solution following your comments: added a `last_demoted_term` field to `struct raft`, and added a test case to ensure that that demoted leader can start the next elections when he is alone.